File: app/src/main/java/com/genonbeta/TrebleShot/app/Activity.java
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package com.genonbeta.TrebleShot.app;

import android.content.*;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.PowerManager;
import android.util.Log;
import android.widget.ImageView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.genonbeta.TrebleShot.App;
import com.genonbeta.TrebleShot.GlideApp;
import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.activity.WelcomeActivity;
import com.genonbeta.TrebleShot.config.AppConfig;
import com.genonbeta.TrebleShot.database.Kuick;
import com.genonbeta.TrebleShot.dialog.ProfileEditorDialog;
import com.genonbeta.TrebleShot.dialog.RationalePermissionRequest;
import com.genonbeta.TrebleShot.object.Identifier;
import com.genonbeta.TrebleShot.object.Identity;
import com.genonbeta.TrebleShot.service.BackgroundService;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableBgTask;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener;
import com.genonbeta.TrebleShot.service.backgroundservice.BackgroundTask;
import com.genonbeta.TrebleShot.service.backgroundservice.BaseAttachableBgTask;
import com.genonbeta.TrebleShot.util.AppUtils;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;

public abstract class Activity extends AppCompatActivity
{
    public static final String
            TAG = Activity.class.getSimpleName(),
            ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED = "android.os.action.POWER_SAVE_MODE_CHANGED";

    public static final int REQUEST_PICK_PROFILE_PHOTO = 1000;

    private final List<BaseAttachableBgTask> mAttachedTaskList = new ArrayList<>();
    private final List<BackgroundTask> mUiTaskList = new ArrayList<>();
    private AlertDialog mOngoingRequest;
    private final IntentFilter mFilter = new IntentFilter();
    private boolean mDarkThemeRequested = false;
    private boolean mAmoledDarkThemeRequested = false;
    private boolean mThemeLoadingFailed = false;
    private boolean mCustomFontsEnabled = false;
    private boolean mSkipPermissionRequest = false;
    private boolean mWelcomePageDisallowed = false;

    private final BroadcastReceiver mReceiver = new BroadcastReceiver()
    {
        @Override
        public void onReceive(Context context, Intent intent)
        {
            if (ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED.equals(intent.getAction()))
                checkForThemeChange();
            else if (BackgroundService.ACTION_TASK_CHANGE.equals(intent.getAction())
                    || App.ACTION_SERVICE_BOUND.equals(intent.getAction()))
                attachTasks();

            Log.d(TAG, "onReceive: " + intent.getAction());
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        mDarkThemeRequested = isDarkThemeRequested();
        mAmoledDarkThemeRequested = isAmoledDarkThemeRequested();
        mCustomFontsEnabled = isUsingCustomFonts();

        mFilter.addAction(ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED);
        mFilter.addAction(BackgroundService.ACTION_TASK_CHANGE);
        mFilter.addAction(App.ACTION_SERVICE_BOUND);

        if (mDarkThemeRequested) {
            try {
                @StyleRes
                int currentThemeRes = getPackageManager().getActivityInfo(getComponentName(), 0).theme;

                Log.d(Activity.class.getSimpleName(), "Activity theme id: " + currentThemeRes);

                if (currentThemeRes == 0)
                    currentThemeRes = getApplicationInfo().theme;

                Log.d(Activity.class.getSimpleName(), "After change theme: " + currentThemeRes);

                @StyleRes
                int appliedRes = 0;

                switch (currentThemeRes) {
                    case R.style.Theme_TrebleShot:
                        appliedRes = R.style.Theme_TrebleShot_Dark;
                        break;
                    case R.style.Theme_TrebleShot_NoActionBar:
                        appliedRes = R.style.Theme_TrebleShot_Dark_NoActionBar;
                        break;
                    case R.style.Theme_TrebleShot_NoActionBar_StaticStatusBar:
                        appliedRes = R.style.Theme_TrebleShot_Dark_NoActionBar_StaticStatusBar;
                        break;
                    default:
                        Log.e(Activity.class.getSimpleName(), "The theme in use for " + getClass().getSimpleName()
                                + " is unknown. To change the theme to what user requested, it needs to be defined.");
                }

                mThemeLoadingFailed = appliedRes == 0;

                if (!mThemeLoadingFailed) {
                    setTheme(appliedRes);

                    if (mAmoledDarkThemeRequested)
                        getTheme().applyStyle(R.style.BlackPatch, true);
                }
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }

        // Apply the Preferred Font Family as a patch if enabled
        if (mCustomFontsEnabled) {
            Log.d(Activity.class.getSimpleName(), "Custom fonts have been applied");
            getTheme().applyStyle(R.style.TextAppearance_Cantarell, true);
        }

        super.onCreate(savedInstanceState);
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        checkForThemeChange();

        if (!hasIntroductionShown() && !mWelcomePageDisallowed) {
            startActivity(new Intent(this, WelcomeActivity.class));
            finish();
        } else if (!AppUtils.checkRunningConditions(this)) {
            if (!mSkipPermissionRequest)
                requestRequiredPermissions(true);
        }
    }

    @Override
    protected void onStart()
    {
        super.onStart();
        attachTasks();
        registerReceiver(mReceiver, mFilter);
        if (AppUtils.checkRunningConditions(this))
            App.notifyActivityInForeground(this, true);
    }

    @Override
    protected void onStop()
    {
        super.onStop();
        unregisterReceiver(mReceiver);
        if (AppUtils.checkRunningConditions(this))
            App.notifyActivityInForeground(this, false);
    }

    @Override
    protected void onDestroy()
    {
        super.onDestroy();
        stopUiTasks();
        detachTasks();
    }

    protected void onAttachTasks(List<BaseAttachableBgTask> taskList)
    {

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults)
    {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (!AppUtils.checkRunningConditions(this))
            requestRequiredPermissions(!mSkipPermissionRequest);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data)
    {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQUEST_PICK_PROFILE_PHOTO)
            if (resultCode == RESULT_OK && data != null) {
                Uri chosenImageUri = data.getData();

                if (chosenImageUri != null) {
                    GlideApp.with(this)
                            .load(chosenImageUri)
                            .centerCrop()
                            .override(200, 200)
                            .into(new CustomTarget<Drawable>()
                            {
                                @Override
                                public void onResourceReady(@NonNull Drawable resource,
                                                            @Nullable Transition<? super Drawable> transition)
                                {
                                    try {
                                        Bitmap bitmap = Bitmap.createBitmap(AppConfig.PHOTO_SCALE_FACTOR,
                                                AppConfig.PHOTO_SCALE_FACTOR, Bitmap.Config.ARGB_8888);
                                        Canvas canvas = new Canvas(bitmap);
                                        FileOutputStream outputStream = openFileOutput("profilePicture",
                                                MODE_PRIVATE);

                                        resource.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
                                        resource.draw(canvas);
                                        bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);

                                        outputStream.close();

                                        notifyUserProfileChanged();
                                    } catch (Exception error) {
                                        error.printStackTrace();
                                    }
                                }

                                @Override
                                public void onLoadCleared(@Nullable Drawable placeholder)
                                {

                                }
                            });
                }
            }
    }

    public void onUserProfileUpdated()
    {

    }

    public void attachTask(BaseAttachableBgTask task)
    {
        synchronized (mAttachedTaskList) {
            if (!mAttachedTaskList.add(task))
                return;
        }

        if (task.getContentIntent() == null)
            task.setContentIntent(this, getIntent());
    }

    private synchronized void attachTasks()
    {
        try {
            BackgroundService service = AppUtils.getBgService(this);
            List<BackgroundTask> concurrentTaskList = service.findTasksBy(getIdentity());
            List<BaseAttachableBgTask> attachableBgTaskList = new ArrayList<>(mAttachedTaskList);
            boolean checkIfExists = attachableBgTaskList.size() > 0;

            // If this call is in between of onStart and onStop, it means there could be some tasks held in the
            // attached task list. To avoid duplicates, we check for them using 'checkIfExists'.
            if (concurrentTaskList.size() > 0) {
                for (BackgroundTask task : concurrentTaskList) {
                    if (task instanceof BaseAttachableBgTask) {
                        BaseAttachableBgTask attachableBgTask = (BaseAttachableBgTask) task;
                        if (!checkIfExists || !attachableBgTaskList.contains(task)) {
                            attachTask(attachableBgTask);
                            attachableBgTaskList.add(attachableBgTask);
                        }
                    }
                }
            }

            // In this phase, we remove the tasks that are no longer known to the background service.
            if (checkIfExists && attachableBgTaskList.size() > 0) {
                if (concurrentTaskList.size() == 0)
                    detachTasks();
                else {
                    List<BaseAttachableBgTask> unrefreshedTaskList = new ArrayList<>(attachableBgTaskList);
                    for (BaseAttachableBgTask task : unrefreshedTaskList) {
                        if (!concurrentTaskList.contains(task))
                            detachTask(task);
                    }
                }
            }

            onAttachTasks(attachableBgTaskList);
            for (BaseAttachableBgTask bgTask : attachableBgTaskList)
                if (!bgTask.hasAnchor())
                    throw new RuntimeException("The task " + bgTask.getClass().getSimpleName() + " owner "
                            + getClass().getSimpleName() + "  did not provide the anchor.");
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    public void attachUiTask(BackgroundTask task)
    {
        synchronized (mUiTaskList) {
            mUiTaskList.add(task);
        }
    }

    public void checkForThemeChange()
    {
        if (((mDarkThemeRequested != isDarkThemeRequested() || (isDarkThemeRequested()
                && mAmoledDarkThemeRequested != isAmoledDarkThemeRequested())) && !mThemeLoadingFailed)
                || mCustomFontsEnabled != isUsingCustomFonts())
            recreate();
    }

    public void checkUiTasks()
    {
        if (mUiTaskList.size() <= 0)
            return;
        synchronized (mUiTaskList) {
            List<BackgroundTask> uiTaskList = new ArrayList<>();
            for (BackgroundTask task : mUiTaskList)
                if (!task.isFinished())
                    uiTaskList.add(task);
            mUiTaskList.clear();
            mUiTaskList.addAll(uiTaskList);
        }
    }

    public void detachTask(BaseAttachableBgTask task)
    {
        synchronized (mAttachedTaskList) {
            task.removeAnchor();
            mAttachedTaskList.remove(task);
        }
    }

    private void detachTasks()
    {
        List<BaseAttachableBgTask> taskList = new ArrayList<>(mAttachedTaskList);
        for (BaseAttachableBgTask task : taskList)
            detachTask(task);
    }

    /**
     * Exits app closing all the active services and connections.
     */
    public void exitApp()
    {
        stopService(new Intent(this, BackgroundService.class));
        finish();
    }

    public List<BaseAttachableBgTask> findTasksWith(Identity identity)
    {
        synchronized (mAttachedTaskList) {
            return BackgroundService.findTasksBy(mAttachedTaskList, identity);
        }
    }

    public Kuick getDatabase()
    {
        return AppUtils.getKuick(this);
    }

    protected SharedPreferences getDefaultPreferences()
    {
        return AppUtils.getDefaultPreferences(this);
    }

    public Identity getIdentity()
    {
        return Identity.withORs(Identifier.from(BackgroundTask.Id.HashCode, BackgroundService.hashIntent(getIntent())));
    }

    public <T extends BaseAttachableBgTask> List<T> getTaskListOf(Class<T> clazz)
    {
        synchronized (mAttachedTaskList) {
            return BackgroundService.getTaskListOf(mAttachedTaskList, clazz);
        }
    }

    public App getSelfApplication()
    {
        return getApplication() instanceof App ? (App) getApplication() : null;
    }

    public boolean hasIntroductionShown()
    {
        return getDefaultPreferences().getBoolean("introduction_shown", false);
    }

    public boolean hasTaskOf(Class<? extends BackgroundTask> clazz)
    {
        synchronized (mAttachedTaskList) {
            return BackgroundService.hasTaskOf(mAttachedTaskList, clazz);
        }
    }

    public boolean hasTaskWith(Identity identity)
    {
        synchronized (mAttachedTaskList) {
            return BackgroundService.hasTaskWith(mAttachedTaskList, identity);
        }
    }

    public boolean isPowerSaveMode()
    {
        if (android.os.Build.VERSION.SDK_INT < 23)
            return false;

        PowerManager powerManager = (PowerManager) getSystemService(Context.POWER_SERVICE);
        return powerManager != null && powerManager.isPowerSaveMode();
    }

    public void interruptAllTasks(boolean userAction)
    {
        if (mAttachedTaskList.size() <= 0)
            return;
        synchronized (mAttachedTaskList) {
            for (BaseAttachableBgTask task : mAttachedTaskList)
                task.interrupt(userAction);
        }
    }

    public boolean isAmoledDarkThemeRequested()
    {
        return getDefaultPreferences().getBoolean("amoled_theme", false);
    }

    public boolean isDarkThemeRequested()
    {
        String value = getDefaultPreferences().getString("theme", "light");
        int systemWideTheme = getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;

        return "dark".equals(value) || ("system".equals(value) && systemWideTheme == Configuration.UI_MODE_NIGHT_YES)
                || ("battery".equals(value) && isPowerSaveMode());
    }

    public boolean isUsingCustomFonts()
    {
        return getDefaultPreferences().getBoolean("custom_fonts", false) && Build.VERSION.SDK_INT >= 16;
    }

    public void loadProfilePictureInto(String deviceName, ImageView imageView)
    {
        try {
            FileInputStream inputStream = openFileInput("profilePicture");
            Bitmap bitmap = BitmapFactory.decodeStream(inputStream);

            GlideApp.with(this)
                    .load(bitmap)
                    .circleCrop()
                    .into(imageView);
        } catch (FileNotFoundException e) {
            imageView.setImageDrawable(AppUtils.getDefaultIconBuilder(this).buildRound(deviceName));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void notifyUserProfileChanged()
    {
        if (!isFinishing())
            runOnUiThread(this::onUserProfileUpdated);
    }

    public void requestProfilePictureChange()
    {
        startActivityForResult(new Intent(Intent.ACTION_PICK).setType("image/*"), REQUEST_PICK_PROFILE_PHOTO);
    }

    public void requestRequiredPermissions(boolean finishIfOtherwise)
    {
        if (mOngoingRequest != null && mOngoingRequest.isShowing())
            return;
        for (RationalePermissionRequest.PermissionRequest request : AppUtils.getRequiredPermissions(this))
            if ((mOngoingRequest = RationalePermissionRequest.requestIfNecessary(this, request,
                    finishIfOtherwise)) != null)
                break;
    }

    public void run(BackgroundTask task)
    {
        task.setContentIntent(this, getIntent());
        BackgroundService.run(this, task);
    }

    /**
     * Run a task that will be stopped if the user leaves.
     *
     * @param task to be stopped when user leaves
     */
    public void runUiTask(BackgroundTask task) {
        attachUiTask(task);
        run(task);
    }

    public <T extends AttachedTaskListener, V extends AttachableBgTask<T>> void runUiTask(V task, T anchor) {
        task.setAnchor(anchor);
        runUiTask(task);
    }

    public void setSkipPermissionRequest(boolean skip)
    {
        mSkipPermissionRequest = skip;
    }

    public void setWelcomePageDisallowed(boolean disallow)
    {
        mWelcomePageDisallowed = disallow;
    }

    public void startProfileEditor()
    {
        new ProfileEditorDialog(this).show();
    }

    protected void stopUiTasks()
    {
        if (mUiTaskList.size() <= 0)
            return;
        synchronized (mUiTaskList) {
            for (BackgroundTask task : mUiTaskList)
                task.interrupt(true);
            mUiTaskList.clear();
        }
    }

    public interface OnBackPressedListener
    {
        boolean onBackPressed();
    }
}

File: app/src/main/java/com/genonbeta/TrebleShot/app/Activity.kt
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.app

import android.content.*
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.os.PowerManager
import android.util.Log
import android.widget.ImageView
import androidx.annotation.StyleRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.genonbeta.TrebleShot.App
import com.genonbeta.TrebleShot.GlideApp
import com.genonbeta.TrebleShot.R
import com.genonbeta.TrebleShot.activity.WelcomeActivity
import com.genonbeta.TrebleShot.config.AppConfig
import com.genonbeta.TrebleShot.database.Kuick
import com.genonbeta.TrebleShot.dataobject.Identifier
import com.genonbeta.TrebleShot.dataobject.Identity
import com.genonbeta.TrebleShot.dataobject.Identity.Companion.withORs
import com.genonbeta.TrebleShot.dialog.ProfileEditorDialog
import com.genonbeta.TrebleShot.dialog.RationalePermissionRequest
import com.genonbeta.TrebleShot.service.BackgroundService
import com.genonbeta.TrebleShot.service.backgroundservice.AsyncTask
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableAsyncTask
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener
import com.genonbeta.TrebleShot.service.backgroundservice.BaseAttachableAsyncTask
import com.genonbeta.TrebleShot.util.AppUtils
import java.io.FileNotFoundException
import java.util.*

abstract class Activity : AppCompatActivity() {
    private var amoledDarkThemeRequested = false

    private val attachedTaskList: MutableList<BaseAttachableAsyncTask> = ArrayList()

    private var customFontsEnabled = false

    private var darkThemeRequested = false

    private val intentFilter = IntentFilter()

    private var ongoingRequest: AlertDialog? = null

    private val receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED == intent.action)
                checkForThemeChange()
            else if (App.ACTION_TASK_CHANGE == intent.action)
                attachTasks()
        }
    }

    lateinit var selfApplication: App
        private set

    var skipPermissionRequest = false
        protected set

    private var themeLoadingFailed = false

    private val uiTaskList: MutableList<AsyncTask> = ArrayList()

    var welcomePageDisallowed = false
        protected set

    override fun onCreate(savedInstanceState: Bundle?) {
        if (application is App)
            selfApplication = application as App
        else
            throw IllegalStateException("This activity cannot run a different Application class.")

        darkThemeRequested = isDarkThemeRequested
        amoledDarkThemeRequested = isAmoledDarkThemeRequested
        customFontsEnabled = isUsingCustomFonts

        intentFilter.addAction(ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED)
        intentFilter.addAction(App.ACTION_TASK_CHANGE)

        if (darkThemeRequested) try {
            @StyleRes var currentThemeRes = packageManager.getActivityInfo(componentName, 0).theme
            Log.d(Activity::class.java.simpleName, "Activity theme id: $currentThemeRes")
            if (currentThemeRes == 0) currentThemeRes = applicationInfo.theme
            Log.d(Activity::class.java.simpleName, "After change theme: $currentThemeRes")
            @StyleRes var appliedRes = 0
            when (currentThemeRes) {
                R.style.Theme_TrebleShot -> appliedRes = R.style.Theme_TrebleShot_Dark
                R.style.Theme_TrebleShot_NoActionBar -> appliedRes = R.style.Theme_TrebleShot_Dark_NoActionBar
                R.style.Theme_TrebleShot_NoActionBar_StaticStatusBar -> appliedRes =
                    R.style.Theme_TrebleShot_Dark_NoActionBar_StaticStatusBar
                else -> Log.e(
                    Activity::class.java.simpleName, "The theme in use for " + javaClass.simpleName
                            + " is unknown. To change the theme to what user requested, it needs to be defined."
                )
            }
            themeLoadingFailed = appliedRes == 0
            if (!themeLoadingFailed) {
                setTheme(appliedRes)
                if (amoledDarkThemeRequested) theme.applyStyle(R.style.BlackPatch, true)
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        // Apply the Preferred Font Family as a patch if enabled
        if (customFontsEnabled) {
            Log.d(Activity::class.java.simpleName, "Custom fonts have been applied")
            theme.applyStyle(R.style.Roundies, true)
        }

        super.onCreate(savedInstanceState)
    }

    override fun onResume() {
        super.onResume()
        checkForThemeChange()
        if (!hasIntroductionShown() && !welcomePageDisallowed) {
            startActivity(Intent(this, WelcomeActivity::class.java))
            finish()
        } else if (!AppUtils.checkRunningConditions(this) && !skipPermissionRequest) {
            requestRequiredPermissions(true)
        }
    }

    override fun onStart() {
        super.onStart()
        attachTasks()
        registerReceiver(receiver, intentFilter)
        selfApplication.notifyActivityInForeground(this, true)
    }

    override fun onStop() {
        super.onStop()
        unregisterReceiver(receiver)
        selfApplication.notifyActivityInForeground(this, false)
    }

    override fun onDestroy() {
        super.onDestroy()
        stopUiTasks()
        detachTasks()
    }

    protected open fun onAttachTasks(taskList: List<BaseAttachableAsyncTask>) {

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (!AppUtils.checkRunningConditions(this))
            requestRequiredPermissions(!skipPermissionRequest)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_PICK_PROFILE_PHOTO) if (resultCode == RESULT_OK && data != null) {
            val chosenImageUri = data.data
            if (chosenImageUri != null) {
                GlideApp.with(this)
                    .load(chosenImageUri)
                    .centerCrop()
                    .override(200, 200)
                    .into(object : CustomTarget<Drawable?>() {
                        override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable?>?) {
                            try {
                                val bitmap = Bitmap.createBitmap(
                                    AppConfig.PHOTO_SCALE_FACTOR,
                                    AppConfig.PHOTO_SCALE_FACTOR,
                                    Bitmap.Config.ARGB_8888
                                )
                                val canvas = Canvas(bitmap)
                                val outputStream = openFileOutput("profilePicture", MODE_PRIVATE)

                                resource.setBounds(0, 0, canvas.width, canvas.height)
                                resource.draw(canvas)
                                bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
                                outputStream.close()
                                notifyUserProfileChanged()
                            } catch (error: Exception) {
                                error.printStackTrace()
                            }
                        }

                        override fun onLoadCleared(placeholder: Drawable?) {}
                    })
            }
        }
    }

    open fun onUserProfileUpdated() {}

    fun attachTask(task: BaseAttachableAsyncTask) {
        synchronized(attachedTaskList) { if (!attachedTaskList.add(task)) return }
        if (task.activityIntent == null)
            task.setContentIntent(this, intent)
    }

    @Synchronized
    private fun attachTasks() {
        val concurrentTaskList = selfApplication.findTasksBy(identity)
        val attachableBgTaskList: MutableList<BaseAttachableAsyncTask> = ArrayList(attachedTaskList)
        val checkIfExists = attachableBgTaskList.size > 0

        // If this call is in between of onStart and onStop, it means there could be some tasks held in the
        // attached task list. To avoid duplicates, we check them using 'checkIfExists'.
        if (concurrentTaskList.isNotEmpty()) {
            for (task in concurrentTaskList) {
                if (task is BaseAttachableAsyncTask) {
                    if (!checkIfExists || !attachableBgTaskList.contains(task)) {
                        attachTask(task)
                        attachableBgTaskList.add(task)
                    }
                }
            }
        }

        // In this phase, we remove the tasks that are no longer known to the background service.
        if (checkIfExists && attachableBgTaskList.size > 0) {
            if (concurrentTaskList.isEmpty())
                detachTasks()
            else {
                val unrefreshedTaskList: List<BaseAttachableAsyncTask> = ArrayList(attachableBgTaskList)
                for (task in unrefreshedTaskList) {
                    if (!concurrentTaskList.contains(task)) detachTask(task)
                }
            }
        }
        onAttachTasks(attachableBgTaskList)
        for (bgTask in attachableBgTaskList) if (!bgTask.hasAnchor()) throw RuntimeException(
            "The task " + bgTask.javaClass.simpleName + " owner "
                    + javaClass.simpleName + " did not provide the anchor."
        )
    }

    fun attachUiTask(task: AsyncTask) {
        synchronized(uiTaskList) { uiTaskList.add(task) }
    }

    fun checkForThemeChange() {
        if ((darkThemeRequested != isDarkThemeRequested || (isDarkThemeRequested
                    && amoledDarkThemeRequested != isAmoledDarkThemeRequested)) && !themeLoadingFailed
            || customFontsEnabled != isUsingCustomFonts
        ) recreate()
    }

    fun checkUiTasks() {
        if (uiTaskList.size <= 0)
            return
        synchronized(uiTaskList) {
            val uiTaskList: MutableList<AsyncTask> = ArrayList()
            for (task in this.uiTaskList) if (!task.finished) uiTaskList.add(task)
            this.uiTaskList.clear()
            this.uiTaskList.addAll(uiTaskList)
        }
    }

    fun detachTask(task: BaseAttachableAsyncTask) {
        synchronized(attachedTaskList) {
            task.removeAnchor()
            attachedTaskList.remove(task)
        }
    }

    private fun detachTasks() {
        val taskList: List<BaseAttachableAsyncTask> = ArrayList(attachedTaskList)
        for (task in taskList) detachTask(task)
    }

    /**
     * Exits app closing all the active services and connections.
     */
    fun exitApp() {
        stopService(Intent(this, BackgroundService::class.java))
        finish()
    }

    fun findTasksWith(identity: Identity): List<BaseAttachableAsyncTask> {
        synchronized(attachedTaskList) {
            return App.findTasksBy(
                attachedTaskList,
                identity
            )
        }
    }

    val database: Kuick
        get() = AppUtils.getKuick(this)

    protected val defaultPreferences: SharedPreferences
        get() = AppUtils.getDefaultPreferences(this)

    open val identity: Identity
        get() = withORs(Identifier.from(AsyncTask.Id.HashCode, AsyncTask.hashIntent(intent)))

    fun <T : BaseAttachableAsyncTask?> getTaskListOf(clazz: Class<T>): List<T> {
        synchronized(attachedTaskList) { return App.getTaskListOf(attachedTaskList, clazz) }
    }

    fun hasIntroductionShown(): Boolean {
        return defaultPreferences.getBoolean("introduction_shown", false)
    }

    fun hasTaskOf(clazz: Class<out AsyncTask?>): Boolean {
        synchronized(attachedTaskList) { return App.hasTaskOf(attachedTaskList, clazz) }
    }

    fun hasTaskWith(identity: Identity): Boolean {
        synchronized(attachedTaskList) { return App.hasTaskWith(attachedTaskList, identity) }
    }

    val isPowerSaveMode: Boolean
        get() {
            if (Build.VERSION.SDK_INT < 23) return false
            val powerManager = getSystemService(POWER_SERVICE) as PowerManager
            return powerManager.isPowerSaveMode
        }

    fun interruptAllTasks(userAction: Boolean) {
        if (attachedTaskList.size <= 0) return
        synchronized(attachedTaskList) { for (task in attachedTaskList) task.interrupt(userAction) }
    }

    val isAmoledDarkThemeRequested: Boolean
        get() = defaultPreferences.getBoolean("amoled_theme", false)

    val isDarkThemeRequested: Boolean
        get() {
            val value = defaultPreferences.getString("theme", "light")
            val systemWideTheme = resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
            return ("dark" == value || "system" == value && systemWideTheme == Configuration.UI_MODE_NIGHT_YES
                    || "battery" == value && isPowerSaveMode)
        }
    val isUsingCustomFonts: Boolean
        get() = defaultPreferences.getBoolean("custom_fonts", false) && Build.VERSION.SDK_INT >= 16

    fun loadProfilePictureInto(deviceName: String, imageView: ImageView) {
        try {
            val inputStream = openFileInput("profilePicture")
            val bitmap = BitmapFactory.decodeStream(inputStream)
            GlideApp.with(this)
                .load(bitmap)
                .circleCrop()
                .into(imageView)
        } catch (e: FileNotFoundException) {
            imageView.setImageDrawable(AppUtils.getDefaultIconBuilder(this).buildRound(deviceName))
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun notifyUserProfileChanged() {
        if (!isFinishing) runOnUiThread { onUserProfileUpdated() }
    }

    fun requestProfilePictureChange() {
        startActivityForResult(Intent(Intent.ACTION_PICK).setType("image/*"), REQUEST_PICK_PROFILE_PHOTO)
    }

    fun requestRequiredPermissions(finishIfOtherwise: Boolean) {
        if (ongoingRequest != null && ongoingRequest!!.isShowing) return
        for (request in AppUtils.getRequiredPermissions(this))
            if (RationalePermissionRequest.requestIfNecessary(this, request, finishIfOtherwise).also {
                    ongoingRequest = it
                } != null
            ) break
    }

    fun run(task: AsyncTask) {
        task.setContentIntent(this, intent)
        App.run(this, task)
    }

    /**
     * Run a task that will be stopped if the user leaves.
     *
     * @param task to be stopped when user leaves
     */
    fun runUiTask(task: AsyncTask) {
        attachUiTask(task)
        run(task)
    }

    fun <T : AttachedTaskListener, V : AttachableAsyncTask<T>> runUiTask(task: V, anchor: T) {
        task.anchor = anchor
        runUiTask(task)
    }

    fun startProfileEditor() {
        ProfileEditorDialog(this).show()
    }

    protected fun stopUiTasks() {
        if (uiTaskList.size <= 0) return
        synchronized(uiTaskList) {
            for (task in uiTaskList) task.interrupt(true)
            uiTaskList.clear()
        }
    }

    interface OnBackPressedListener {
        fun onBackPressed(): Boolean
    }

    companion object {
        val TAG = Activity::class.java.simpleName
        const val ACTION_SYSTEM_POWER_SAVE_MODE_CHANGED = "android.os.action.POWER_SAVE_MODE_CHANGED"
        const val REQUEST_PICK_PROFILE_PHOTO = 1000
    }
}